proposals/NNNN-uniformity-qualifiers.md

```diff
@@ -0,0 +1,132 @@
+<!-- {% raw %} -->
+
+# Uniformity Qualifiers
+
+* Proposal: [NNNN](NNNN-uniformity-qualifiers.md)
+* Author(s): [Chris Bieneman](https://github.com/llvm-beanz)
+* Sponsor: [Chris Bieneman](https://github.com/llvm-beanz)
+
+* Status: **Under Consideration**
+
+## Introduction
+
+The HLSL Single Program Multiple Data (SPMD) programming model defines a program
+in terms of how it operates on a single element of data. An SPMD program may be
+executed on a traditional scalar processor or a Single Instruction Multiple Data
+(SIMD) processor. When executing on a SIMD processor the program may execute
+where a single instruction produces results for multiple threads of execution
+from the source programming model. This is sometimes referred to as Single
+Instruction Multiple Threads (SIMT).
+
+Under HLSL's execution model, groups of threads form hierarchical scopes:
+* A _dispatch_ represents the full set of threads spawned from a CPU API
+  invocation.
+* A _thread group_ represents a subset of a dispatch that can execute
+  concurrently.
+* A _wave_ represents a subset of a thread group that represents in a single
+  SIMD processor.
+* A _quad_ represents a grouping of four adjacent threads in a wave.
+
+When a shader is executing threads concurrently on one or more processing cores,
+an emergent property of _uniformity_ exists within the thread group, wave and
+quad scopes.
+
+Uniformity can refer to data or control flow. If a variable has the same value
+across all threads in a scope, it is said to be _uniform_ across that scope.
+Similarly if all threads within a scope are actively executing instructions
+within a control flow block, the control flow is said to be _uniform control
+flow_ across that scope.
+
+## Motivation
+
+Uniformity of data and control flow are central concepts to SIMT execution
+models, and is required for correct execution of shader programs. Despite
+the importance of this fundamental property it is not represented in any
+explicit way in the HLSL language.
+
+This proposal, seeks to address that by introducing core concepts around
+uniformity to HLSL's type system and programming model.
+
+## Proposed solution
+
+### Uniformity as a Type Qualifier
+
+This proposal introduces a new set of type qualifiers to represent the different
+scopes of uniformity:
+* `group_uniform`
+* `simd_uniform`
+* `quad_uniform`
+* non-uniform (default state with no associated keyword)
+
+`group_uniform` is the highest scoping of uniformity, and implies all other
+scopes. `simd_uniform` implies `quad_uniform`.
+
+A new "UniformityReduction" cast will reduce the uniform scope allowing
+conversion of one uniformity scope to another uniformity scope as long as the
+source scope has a greater uniformity scope.

```

Comment:
Something i'm missing here is what happens if you did, e.g.:

```hlsl
group_uniform int a;
simd_uniform int b;
int c = a * b;
```

I would _expect_ c to automatically be treated as simd_uniform (i.e. the smaller of the scopes) as the result, but that's not spelled out here AFAICT.

Reply:
Oh wait now I look again that's what the next two sentences are saying - perhaps worth an example?